witchcraft-server/src/service/gzip.rs

```diff
@@ -188,57 +188,74 @@ where
 
     type Error = B::Error;
 
-    fn poll_data(
+    fn poll_frame(
         self: Pin<&mut Self>,
         cx: &mut Context<'_>,
-    ) -> Poll<Option<Result<Self::Data, Self::Error>>> {
+    ) -> Poll<Option<Result<Frame<Self::Data>, Self::Error>>> {
         let mut this = self.project();
 
-        if *this.done {
-            return Poll::Ready(None);
-        }
-
-        let next = ready!(this.body.as_mut().poll_data(cx)).transpose()?;
-
-        let Some(encoder) = this.encoder else {
-            return Poll::Ready(next.map(Ok));
-        };
-
-        match next {
-            Some(next) => {
-                encoder.write_all(&next).unwrap();
-                if this.body.is_end_stream() {
+        match mem::replace(this.state, State::Done) {
+            State::Compressing(mut encoder) => match this.body.as_mut().poll_frame(cx) {
+                Poll::Ready(Some(Ok(frame))) => match frame.data_ref() {
+                    Some(data) => {
+                        encoder.write_all(data).unwrap();
+                        if this.body.is_end_stream() {
+                            encoder.try_finish().unwrap();
+                            let buf = encoder.get_mut().get_mut().split().freeze();
+                            Poll::Ready(Some(Ok(Frame::data(buf))))
+                        } else {
+                            encoder.flush().unwrap();
+                            let buf = encoder.get_mut().get_mut().split().freeze();
+                            *this.state = State::Compressing(encoder);
+                            Poll::Ready(Some(Ok(Frame::data(buf))))
+                        }
```

Comment:
nit: maybe do this after the if/else?

```
// assuming you don't need to do this before re-assigning this.sate in the else
let buf = encoder.get_mut().get_mut().split().freeze();
Poll::Ready(Some(Ok(Frame::data(buf))))
```

Reply:
We do need to construct the buf before reassigning the state, yeah since it holds the encoder.